### PR TITLE
✨ Error creation

### PR DESCRIPTION
**Pull request by kirrg001**
refs #7116
### overview changes
- remove errors folder
- option pattern for errors
- main Error GhostError - inheritance
- always use `Error` postfix naming
- express error-handler takes care that each error get's wrapped into a GhostError if not happened yet
- add level (`critical`, `normal`)
  - normal is default -> this is very helpful
  - for example 404 is not critical, but database version error or scheduling error is
  - so it's very easy to filter them later.
### GhostError and inheritance

Before this PR, each error had no prototype to inherit from.

That has some disadvantages:
- you always need to define all properties an error has
- can happen that some errors are inconsistent to other errors
- no default set of attribute values
- we had several un-needed checks
- context/help was not a default property
- we were not able to capture the error stack from the original error (now possible by just passing the original error) (https://github.com/TryGhost/Ghost/pull/7477/files#diff-25f0b2b0cc338d73a55da195594de403R278)

By default the `GhostError` prototype is an `InternalServerError` with statusCode 500. We can change that behaviour if we want.
### Option pattern for errors

We always passed error arguments in a row (`new errors.IncorrectUsageError(a,b,c)`).
This has a big disadvantage: if you want to add or change the arguments order, you need to change everything. With option pattern you can pass how much arguments you want and it's even more readable for the caller/developer (`new errors.IncorrectUsageError(true, 'hello')` vs `new errors.IncorrectUsageError({hideStack: true, message: 'hello'})`).
### Error handling

This is not particular part of this PR, but we should take care that units return a proper Ghost error instance. So for example if you create a unit A, which uses a third party module, which returns errors as pure strings (yes that happens sometimes), then it's your responsibility to react on these error strings and return a Ghost error instance instead. So checks like https://github.com/TryGhost/Ghost/pull/7477/files#diff-517787c62b3fa7bfcac648c0763b07ecL41 won't happen in the future. We had 100 of these use cases in our old error handler file, which were 1. hard to understand and 2. un-needed, when we do proper error creation.

@ErisDS 
https://github.com/TryGhost/Ghost/pull/7477/files#diff-7a5f479230bbe115458d26c3c7026164L94
I deleted the `debug` feature, do you want to keep it and if yes, can you explain why?:)

https://github.com/TryGhost/Ghost/pull/7477/commits/d6b91266e4aca0f028c18069548dec9fb8a474d2
This is a separate commit. I can extract this change if you want 🌮

### CODE REVIEW

File: content/themes/casper [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 95cec988f5b0dc0d473ab11222fe78105517225f
+Subproject commit 5c51195ca6cafbe3948c72b1461ad91b6195cff3
```
Comment (ErisDS) [inline content/themes/casper]: I saw it, so I thought I'd flag it 😉

**Comment by kirrg001** [comment]
ready for review 🇹🇭

File: core/server/config/index.js [hunk 2]
```diff
@@ -38,6 +36,7 @@ localUtils.makePathsAbsolute.bind(nconf)();
  * @TODO: ghost-cli?
  */
 nconf.set('ghostVersion', packageInfo.version);
+nconf.set('env', env);
```
Comment (ErisDS) [inline core/server/config/index.js]: 👍

File: core/server/errors.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    Error.call(this);
+    Error.captureStackTrace(this, GhostError);
+
+    /**
+     * defaults
+     */
+    this.statusCode = 500;
+    this.errorType = 'InternalServerError';
+    this.level = 'normal';
```
Comment (ErisDS) [inline core/server/errors.js]: TODO Note: I'd like to add the usage of an individual ID to errors, as we have in ignition

File: core/server/middleware/error-handler.js [hunk 4]
```diff
@@ -88,6 +89,12 @@ module.exports = function errorHandler(err, req, res, next) {
         err = err[0];
     }
 
+    if (!(err instanceof errors.GhostError)) {
+        err = new errors.GhostError({
+            err: err
+        });
```
Comment (ErisDS) [inline core/server/middleware/error-handler.js]: One question I have here is how does this impact the stack trace?

Comment (ErisDS) [inline core/server/middleware/error-handler.js]: Did a quick test (deleted core/server/views/default.hbs) and answered my own question. This works great 👍 🎉

**Review by ErisDS — COMMENTED** [review]
> @ErisDS 
> https://github.com/TryGhost/Ghost/pull/7477/files#diff-7a5f479230bbe115458d26c3c7026164L94
> I deleted the debug feature, do you want to keep it and if yes, can you explain why? :)

Debug shows me the number of milliseconds since the last debug call. Therefore, as I have debug statements inside of the controllers, this piece of (definitely ugly) code means I can see what is taking the time between when the request is received and when it is responded to in a single interface. I'd like to keep this until I have finished the refactoring work I'm doing, i.e. until the end of the alpha

> d6b9126
> This is a separate commit. I can extract this change if you want 🌮

The commit doesn't exist, not sure what it was?

File: core/server/errors.js [hunk 5]
```diff
@@ -13,6 +13,7 @@ function GhostError(options) {
 
     /**
      * defaults
+     * @TODO: I'd like to add the usage of an individual ID to errors, as we have in ignition
```
Comment (kirrg001) [inline core/server/errors.js]: added a TODO as well into #7116

File: core/server/middleware/index.js [hunk 6]
```diff
@@ -92,6 +92,17 @@ setupMiddleware = function setupMiddleware(blogApp) {
         next();
     });
 
+    if (debug.enabled) {
```
Comment (kirrg001) [inline core/server/middleware/index.js]: reverted 🎉

**Comment by ErisDS** [comment]
This PR wins a :cake: and a 🍾  because we can now see errors for what is causing the random Travis failures 🎉  

I've logged the output to #7470 to deal with later, and restarted the builds so we can merge 👯

**Comment by ErisDS** [comment]
Wow, it took me 6 restarts on node v6+MySQL to get this to go green 😲  something to look into next week.

**Pull request merged**